ci/appveyor-cpp-build.bat

```diff
@@ -17,114 +17,146 @@
-)
+@rem Use Boost from Anaconda
+set BOOST_ROOT=%CONDA_PREFIX%\Library
+set BOOST_LIBRARYDIR=%CONDA_PREFIX%\Library\lib
+
+@rem The "main" C++ build script for Windows CI
+@rem (i.e. for usual configurations)
 
 if "%JOB%" == "Toolchain" (
-  @rem Install pre-built "toolchain" packages for faster builds
-  set CONDA_PACKAGES=%CONDA_PACKAGES% --file=ci\conda_env_cpp.yml
+  set CMAKE_ARGS=-DARROW_DEPENDENCY_SOURCE=CONDA -DARROW_WITH_BZ2=ON
+) else (
+  @rem We're in a conda environment but don't want to use it for the dependencies
+  set CMAKE_ARGS=-DARROW_DEPENDENCY_SOURCE=AUTO
 )
 
-conda create -n arrow -q -y %CONDA_PACKAGES% -c conda-forge || exit /B
+@rem Enable warnings-as-errors
+set ARROW_CXXFLAGS=/WX /MP
 
-call activate arrow
+@rem
+@rem Build and test Arrow C++ libraries (including Parquet)
+@rem
 
-@rem Use Boost from Anaconda
-set BOOST_ROOT=%CONDA_PREFIX%\Library
-set BOOST_LIBRARYDIR=%CONDA_PREFIX%\Library\lib
+mkdir cpp\build
+pushd cpp\build
+
+@rem XXX Without forcing CMAKE_CXX_COMPILER, CMake can re-run itself and
+@rem unfortunately switch from Release to Debug mode...
+@rem
+@rem In release mode, disable optimizations (/Od) for faster compiling
+@rem and enable runtime assertions.
+
+cmake -G "%GENERATOR%" %CMAKE_ARGS% ^
+      -DARROW_BOOST_USE_SHARED=OFF ^
+      -DARROW_BUILD_EXAMPLES=ON ^
+      -DARROW_BUILD_STATIC=OFF ^
+      -DARROW_BUILD_TESTS=ON ^
+      -DARROW_CSV=ON ^
+      -DARROW_CXXFLAGS="%ARROW_CXXFLAGS%" ^
+      -DARROW_DATASET=ON ^
+      -DARROW_ENABLE_TIMING_TESTS=OFF ^
+      -DARROW_FLIGHT=%ARROW_BUILD_FLIGHT% ^
+      -DARROW_GANDIVA=%ARROW_BUILD_GANDIVA% ^
+      -DARROW_MIMALLOC=ON ^
+      -DARROW_PARQUET=ON ^
+      -DARROW_PYTHON=ON ^
+      -DARROW_S3=%ARROW_S3% ^
+      -DARROW_VERBOSE_THIRDPARTY_BUILD=OFF ^
+      -DARROW_WITH_BROTLI=ON ^
+      -DARROW_WITH_LZ4=ON ^
+      -DARROW_WITH_SNAPPY=ON ^
+      -DARROW_WITH_ZLIB=ON ^
+      -DARROW_WITH_ZSTD=ON ^
+      -DCMAKE_BUILD_TYPE="Release" ^
+      -DCMAKE_CXX_COMPILER=clcache ^
+      -DCMAKE_CXX_FLAGS_RELEASE="/MD /Od /UNDEBUG" ^
+      -DCMAKE_INSTALL_PREFIX=%CONDA_PREFIX%\Library ^
+      -DCMAKE_UNITY_BUILD=ON ^
+      -DCMAKE_VERBOSE_MAKEFILE=OFF ^
+      -DPARQUET_BUILD_EXECUTABLES=ON ^
+      -DPARQUET_REQUIRE_ENCRYPTION=ON ^
+      ..  || exit /B
+cmake --build . --target install --config %CONFIGURATION%  || exit /B
+
+@rem Needed so arrow-python-test.exe works
+set OLD_PYTHONHOME=%PYTHONHOME%
+set PYTHONHOME=%CONDA_PREFIX%
+
+ctest --output-on-failure -j2 || exit /B
+
+set PYTHONHOME=%OLD_PYTHONHOME%
+popd
+
+@rem
+@rem Build and install pyarrow
+@rem
+
+pushd python
+
+set PYARROW_BUNDLE_BOOST=OFF
+set PYARROW_CMAKE_GENERATOR=%GENERATOR%
+set PYARROW_CXXFLAGS=%ARROW_CXXFLAGS%
+set PYARROW_PARALLEL=2
+set PYARROW_WITH_DATASET=ON
+set PYARROW_WITH_FLIGHT=%ARROW_BUILD_FLIGHT%
+set PYARROW_WITH_GANDIVA=%ARROW_BUILD_GANDIVA%
+set PYARROW_WITH_PARQUET=ON
+set PYARROW_WITH_S3=%ARROW_S3%
+set PYARROW_WITH_STATIC_BOOST=ON
```

Comment:
I'm not sure the Boost-related settings are still useful?

Reply:
I'm takling that #7126